builtin/logical/totp/path_keys.go

```diff
@@ -0,0 +1,400 @@
+			},
+
+			"url": {
+				Type:        framework.TypeString,
+				Description: `A TOTP url string containing all of the parameters for key setup. Only used if generate is false.`,
+			},
+		},
+
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.ReadOperation:   b.pathKeyRead,
+			logical.UpdateOperation: b.pathKeyCreate,
+			logical.DeleteOperation: b.pathKeyDelete,
+		},
+
+		HelpSynopsis:    pathKeyHelpSyn,
+		HelpDescription: pathKeyHelpDesc,
+	}
+}
+
+func (b *backend) Key(s logical.Storage, n string) (*keyEntry, error) {
+	entry, err := s.Get("key/" + n)
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return nil, nil
+	}
+
+	var result keyEntry
+	if err := entry.DecodeJSON(&result); err != nil {
+		return nil, err
+	}
+
+	return &result, nil
+}
+
+func (b *backend) pathKeyDelete(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	err := req.Storage.Delete("key/" + data.Get("name").(string))
+	if err != nil {
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+func (b *backend) pathKeyRead(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	key, err := b.Key(req.Storage, data.Get("name").(string))
+	if err != nil {
+		return nil, err
+	}
+	if key == nil {
+		return nil, nil
+	}
+
+	// Translate algorithm back to string
+	algorithm := key.Algorithm.String()
+
+	// Return values of key
+	return &logical.Response{
+		Data: map[string]interface{}{
+			"issuer":       key.Issuer,
+			"account_name": key.AccountName,
+			"period":       key.Period,
+			"algorithm":    algorithm,
+			"digits":       key.Digits,
+		},
+	}, nil
+}
+
+func (b *backend) pathKeyList(
+	req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	entries, err := req.Storage.List("key/")
+	if err != nil {
+		return nil, err
+	}
+
+	return logical.ListResponse(entries), nil
+}
+
+func (b *backend) pathKeyCreate(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	name := data.Get("name").(string)
+	generate := data.Get("generate").(bool)
+	keyString := data.Get("key").(string)
+	issuer := data.Get("issuer").(string)
+	accountName := data.Get("account_name").(string)
+	period := data.Get("period").(int)
+	algorithm := data.Get("algorithm").(string)
+	digits := data.Get("digits").(int)
+	skew := data.Get("skew").(int)
+	qrSize := data.Get("qr_size").(int)
+	keySize := data.Get("key_size").(int)
+	inputURL := data.Get("url").(string)
+
+	if generate == true {
+		if keyString != "" {
+			return logical.ErrorResponse("a key should not be passed if generate is true"), nil
+		}
+		if inputURL != "" {
+			return logical.ErrorResponse("a url should not be passed if generate is true"), nil
+		}
+	}
+
+	// Read parameters from url if given
+	if inputURL != "" {
+		//Parse url
+		urlObject, err := url.Parse(inputURL)
+		if err != nil {
+			return logical.ErrorResponse("an error occured while parsing url string"), nil
```

Comment:
I did mean log the error, although I guess depending on what that error is, if it spits out parts of the url it could log something secret. It may be worth returning the error to the caller though, presumably since they provided the URL they can be privy to parts of it in the error message.